### DISCUSSION TITLE
Browser noticably more sluggish than "desktop"

### DISCUSSION THREAD

**Discussion by houstonhaynes** [JordanMarr/ReactiveElmish.Avalonia#18] [General]
Hey I was just noodling around with the "browser" output of XPlat and noticed that the responsiveness to on-screen actions was SEVERELY laggy. The difference between that and the "desktop" apps is *not* trivial. Is this a known Avalonia issue or is this something "local" to Elmish?

**Comment by JordanMarr** [comment]
I wonder if this is due to the fact that there are multiple Elmish loops running (since there is one for each page).
Trying out your latest PR, I noticed that the chart subscription still runs when tabbing to another page. 

While it may be desirable in some cases to continue running a loop when a page is inactive, I think may need to add the ability to dispose an Elmish loop when a page/view is no longer active.

**Comment by JordanMarr** [comment]
I made an experimental branch to try out a temporary workaround to ensure that the Chart auto-update is false to deactivate the subscription.
Basically, the `MainViewModel` sends out a new `TabChanged` event when a tab is changed.
The `ChartViewModel` subscribes to that event and then disables the auto-update (which then disposes the timer subscription).

I don't love this approach becuase it feels very manual; but it does work.
I think the better approach will be for the framework to provide a built-in way for VM elmish loops to unsubscribe themselves.

https://github.com/JordanMarr/Elmish.Avalonia/commit/14ebe4f6c5ef86bf8ca2c0642acaaa93973d88eb

**Comment by houstonhaynes** [comment]
Really interesting question - I wondered to myself if clicking on "Ok" in the Chart view would dispose of the subscription.

But with "I noticed that the chart subscription still runs when tabbing to another page." I was wrong in thinking that the loop is *quiet* until IsAutoUpdating is set to true (and in my note in another thread I pondered that setting IsAutoUpdating = false as a way of 'switching it off' again)

**Comment by JordanMarr** [comment]
Still experimenting, but I think this is starting to go down the right path:
https://github.com/JordanMarr/Elmish.Avalonia/tree/loop-terminate

It adds an optional `stopLoopWhenViewIsHidden` parameter to the `ElmishViewModel`. 

```F#
let vm = 
    ElmishViewModel(
        AvaloniaProgram.mkSimple init update bindings
        |> AvaloniaProgram.withSubscription subscriptions
        , stopLoopWhenViewIsHidden = true
    )
```

`stopLoopWhenViewIsHidden ` defaults to `false` at the moment, but maybe it should default to `true` because the `ViewLocator` automatically initializes a new loop every time it sees an `IElmishViewModel` referenced in a view (every time a tab is opened).
(OTOH, using a `ViewLocator` is totally optional. It is possible to manually bind an `IElmishViewModel` to a view.)

Once I get the API nailed down, I will include `ElmishViewModel` into the `Elmish.Avalonia` library.

**Comment by JordanMarr** [comment]
Found some issues with this approach, but I think I have another idea to try that will be better.

**Reply by houstonhaynes** [reply]
That's great. I had wondered whether "shutting down" non-visible elements of the app would have adverse consequences in certain cases. 

I also wondered if the "threaded model" in WASM made for thread starving that wouldn't be (necessarily) present in native.

**Comment by daz10000** [comment]
On Fri, Aug 04, 2023 at 04:20:53AM -0700, Houston Haynes wrote:
> That's great. I had wondered whether "shutting down" non-visible elements of the app would have adverse consequences in certain cases. 
> 
> I also wondered if the "threaded model" in WASM made for thread starving that wouldn't be (necessarily) present in native.
> 
> 

I have noticed in general that the wasm / web version isn't as snappy as the desktop UI.  I put it down to just wasm being overall slower running dotnet, but if there are some background tasks that aren't necessary and can be disabled, that would be a nice improvement.  The desktop versions are so responsive, it's refreshing

**Comment by JordanMarr** [comment]
I'm feeling pretty good about this implementation.
It looks like this:

[ChartViewModel.fs](https://github.com/JordanMarr/Elmish.Avalonia/blob/d63f3e2d86a2e2e3254920e926c0aa097a6bc2ae/src/Samples/AvaloniaExample/ViewModels/ChartViewModel.fs#L172)
```F#
let vm = 
    AvaloniaProgram.mkSimple init update bindings
    |> AvaloniaProgram.withSubscription subscriptions
    |> ElmishViewModel.create
    |> ElmishViewModel.terminateOnUnloaded Terminate
```

## Changes
* Added a new `ElmishViewModel` module that is more pipe friendly
* Moved ElmishViewModel.fs to the Elmish.Avalonia project so it will be available in the upcoming NuGet version (rather than part of the template)
* Added `ElmishViewModel.terminateOnUnloaded`. This requires that you create a `Terminate` message. (You can call it whatever you want; its only purpose is to trigger loop termination.) Behind the scenes, this does the following:
  * Appends a subscription based on the view `Unloaded` event that dispatches the passed-in `Terminate` message.
  * Wires up the Elmish `Program.withTermination` to terminate the loop when the `Terminate` message is dispatched.

One thing I'm undecided on is whether `terminateOnUnloaded` should just happen by default. 
I think maybe it should.


It's in the "loop-terminate" branch if you want to mess with it.

**Comment by JordanMarr** [comment]
[Elmish.Avalonia v1.0.1](https://github.com/JordanMarr/Elmish.Avalonia/releases/tag/1.0.1) is released!

[ElmishViewModel Docs](https://github.com/JordanMarr/Elmish.Avalonia#elmishviewmodel)

**Reply by houstonhaynes** [reply]
That's really great. I have to admit that I'm a little confused about **subscription loops** running in background (not disposed of even when the user switched away to another view) versus the MVU loop *itself* chugging away behind the scenes. 

I'm thinking about those cases where something like an external event comes in to trigger a command/message. (I'm thinking of something like a BLE characteristic changing on a server that notifies some element of the app which pertains to a view that's **not** in focus) 

Am I missing a part of the picture here?

**Reply by JordanMarr** [reply]
The Elmish loop fires off the subscription function on each iteration, so terminating the Elmish loop also disposes your subscriptions. 

The new `terminateOnUnloaded` allows you to automatically terminate your loop and your subscriptions when the view unloads. 

In the sample app, clicking to another tab unloads the current tab’s view, which terminates the loop and subscriptions. 

Does that clear up the confusion?

**Reply by houstonhaynes** [reply]
Yes - and I'm not sure that's really desirable - though now I think about it - **if** a user returns to that view by clicking on the menu item at the top of the MainView then it would kick off init with a fresh model anyway...

So let me ask a hypothetical... 

Let's just say that the user is on a screen where they're doing something with WiFi settings to send to an IoT device via Bluetooth. So the user is in a WiFi settings "view" and nothing to do with a view that has something to do with display of - say - the angle of a solar panel using an accelerometer. While that user is busy in another view, something happens when the solar panel is being "moved" by the IoT device and an error occurs. That's the kind of error that should *always* surface to the user regardless of 1) which view they're in or 2) whether they're actively in the app or not...

Where does that over-riding "master state" function *live* in this kind of app?

I ask this because I could imagine that an MVU loop where the functionality that pertains to managing/monitoring the angle of the solar panel would be the place where that would surface - which *also* means that the return of the user to that view would **not** re-trigger a new model state- but rather pick up and continue with the model that was there at the time the user moved off to another view.

Do you understand the scenario? What should be expected when the user wishes to have this behavior "off-screen"? Background service?

**Reply by JordanMarr** [reply]
In that case, you would make sure that your secondary view was not unloaded, or you would take special care to unload it only when you’re ready. 

The sample app just unloads the current view when you go to a different tab, but it doesn’t have to be that way. If it used the avalonia tab control, I’m pretty sure the view would not unload; it would just be hidden, so the loop would continue running.

**Reply by houstonhaynes** [reply]
Right on. That's really helpful to know. 👍

**Comment by houstonhaynes** [comment]
So I'll ask one last "you should probably know this already" question - does the WASM browser instance run single-threaded or is it multi-threaded? I know they had talked multi-threading WASM for .NET 7 but not sure where they landed on it. At the time those threads (pun!) were going around I was dealing with film studio DRM so the WASM sandbox was out of the question for me either way.

This makes me wonder whether there *should* be a setting to auto-magically shut down all loops in WASM *other* than the active view's MVU loop... at least if it's still running single-threaded.

**Reply by JordanMarr** [reply]
I don’t know about the WASM thread question, but I don’t think i can automatically terminate a loop anyways because Elmish requires you create  a message to terminate the loop, so it has to be done manually.

**Reply by houstonhaynes** [reply]
It just made me wonder that - even if doing a "good" job minding which views were active-vs-terminated that a single-threaded WASM instance may bog down as the various processes contend for that thread. It's something I'll peruse over the next week when I return to civilization. 🏕️ 😄

**Comment by daz10000** [comment]
Seeing the web version again, I'm still noticing lag a lot and it seems to be different between the two apps.  The counter app is surprisingly one of the slowest.   If you hit the + button many times in a row, it clearly registers the clicks, but it's almost a second between updates (suspiciously close to a second tbh, like something is polling every now and then or there's a delay built in).  Harder to tell on the other apps, but they seem more responsive.  I know wasm will be slow compared to desktop (which is really responsive after years of using slow web apps in general), given how it's running under the hood, but this feels like something else on top of the slower performance.   I will keep an eye on it - I have no idea how you'd profile a dotnet app running on wasm :) but there might be way.

**Comment by daz10000** [comment]
This time chrome is the star (I try not to use it but [this article recommended the performance tab](https://remibou.github.io/Profiling-in-Blazor/))
It produces a nice flame graph.  Ironically the app itself is way more responsive as well ( this is taking the code base post PR merge), so something else was probably fixed.. It feels a lot smoother, even in FF now.  cest la vie.  Not sure how helpful the browser profiling will be without annotating the actual functions - print statements inside the code might be more informative.  It's all about the mono runtime from what I can see and wasm/js interactions.  The longest delays seem to be drawing / animation frame related

![image](https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/6342645/0305350b-a5ad-4a48-a196-9f1ba49601c1)


![image](https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/6342645/8ab23037-9615-4322-acba-08a8a527cc3a)

**Comment by JordanMarr** [comment]
There was an extra call to `OnFrameworkInitializationCompleted()` that i found and deleted here:
https://github.com/JordanMarr/ReactiveElmish.Avalonia/commit/85a6e1a1bf8dd1b14b22bd0feb3acdb9e6f0174c

Maybe that had something to do with it?
